src/libmemcached/man/memcached_set.3
.TH "MEMCACHED_SET" "3" "April 03, 2013" "1.0.17" "libmemcached"
.SH NAME
memcached_set \- Storing and Replacing Data
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructuredText.
.
.SH SYNOPSIS
.sp
#include <libmemcached/memcached.h>
.INDENT 0.0
.TP
.B memcached_return_t memcached_set(memcached_st\fI\ *ptr\fP, const char\fI\ *key\fP, size_t\fI\ key_length\fP, const char\fI\ *value\fP, size_t\fI\ value_length\fP, time_t\fI\ expiration\fP, uint32_t\fI\ flags\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_add(memcached_st\fI\ *ptr\fP, const char\fI\ *key\fP, size_t\fI\ key_length\fP, const char\fI\ *value\fP, size_t\fI\ value_length\fP, time_t\fI\ expiration\fP, uint32_t\fI\ flags\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_replace(memcached_st\fI\ *ptr\fP, const char\fI\ *key\fP, size_t\fI\ key_length\fP, const char\fI\ *value\fP, size_t\fI\ value_length\fP, time_t\fI\ expiration\fP, uint32_t\fI\ flags\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_set_by_key(memcached_st\fI\ *ptr\fP, const char\fI\ *group_key\fP, size_t\fI\ group_key_length\fP, const char\fI\ *key\fP, size_t\fI\ key_length\fP, const char\fI\ *value\fP, size_t\fI\ value_length\fP, time_t\fI\ expiration\fP, uint32_t\fI\ flags\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_add_by_key(memcached_st\fI\ *ptr\fP, const char\fI\ *group_key\fP, size_t\fI\ group_key_length\fP, const char\fI\ *key\fP, size_t\fI\ key_length\fP, const char\fI\ *value\fP, size_t\fI\ value_length\fP, time_t\fI\ expiration\fP, uint32_t\fI\ flags\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_replace_by_key(memcached_st\fI\ *ptr\fP, const char\fI\ *group_key\fP, size_t\fI\ group_key_length\fP, const char\fI\ *key\fP, size_t\fI\ key_length\fP, const char\fI\ *value\fP, size_t\fI\ value_length\fP, time_t\fI\ expiration\fP, uint32_t\fI\ flags\fP)
.UNINDENT
.sp
Compile and link with \-lmemcached
.SH DESCRIPTION
.sp
\fI\%memcached_set()\fP, \fI\%memcached_add()\fP, and \fI\%memcached_replace()\fP are all used to store information on the server. All methods take a key, and its length to store the object. Keys are currently limited to 250 characters when using either a version of memcached(1) which is 1.4 or below, or when using the text protocol. You must supply both a value and a length. Optionally you
store the object. Keys are currently limited to 250 characters by the
memcached(1) server. You must supply both a value and a length. Optionally you
may test an expiration time for the object and a 16 byte value (it is meant to be used as a bitmap). "flags" is a 4byte space that is stored alongside of the main value. Many sub libraries make use of this field, so in most cases users should avoid making use of it.
.sp
\fI\%memcached_set()\fP will write an object to the server. If an object
already exists it will overwrite what is in the server. If the object does not
exist it will be written. If you are using the non\-blocking mode this function
will always return true unless a network error occurs.
.sp
\fI\%memcached_replace()\fP replaces an object on the server. If the object is not found on the server an error occurs.
.sp
\fI\%memcached_add()\fP adds an object to the server. If the object is found on the server an error occurs, otherwise the value is stored.
.sp
\fBmemcached_cas()\fP overwrites data in the server as long as the "cas"
value is still the same in the server. You can get the cas value of a result by
calling \fBmemcached_result_cas()\fP on a memcached_result_st(3) structure. At the point
that this note was written cas is still buggy in memached. Turning on tests
for it in libmemcached(3) is optional. Please see memcached_set for
information on how to do this.
.sp
\fI\%memcached_set_by_key()\fP, \fI\%memcached_add_by_key()\fP, and \fI\%memcached_replace_by_key()\fP methods all behave in a similar method as the non
key methods. The difference is that they use their group_key parameter to map
objects to particular servers.
.sp
If you are looking for performance, \fI\%memcached_set()\fP with non\-blocking IO is the fastest way to store data on the server.
.sp
All of the above functions are testsed with the \fBMEMCACHED_BEHAVIOR_USE_UDP\fP behavior enabled. However, when using these operations with this behavior
on, there are limits to the size of the payload being sent to the server.
The reason for these limits is that the Memcached Server does not allow
multi\-datagram requests and the current server implementation sets a datagram
size to 1400 bytes. Due to protocol overhead, the actual limit of the user supplied data is less than 1400 bytes and depends on the protocol in use as, well as the operation being
executed. When running with the binary protocol, \fBMEMCACHED_BEHAVIOR_BINARY_PROTOCOL\fP, the size of the key,value, flags and expiry combined may not
exceed 1368 bytes. When running with the ASCII protocol, the exact limit fluctuates depending on which function is being executed and whether the function is a cas operation or not. For non\-cas ASCII set operations, there are at least
1335 bytes available to split among the key, key_prefix, and value; for cas
ASCII operations there are at least 1318 bytes available to split among the key, key_prefix and value. If the total size of the command, including overhead,
exceeds 1400 bytes, a \fBMEMCACHED_WRITE_FAILURE\fP will be returned.
.SH RETURN
.sp
All methods return a value of type \fBmemcached_return_t\fP.
On success the value will be \fBMEMCACHED_SUCCESS\fP.
Use \fBmemcached_strerror()\fP to translate this value to a printable string.
.sp
For \fI\%memcached_replace()\fP and \fI\%memcached_add()\fP, \fBMEMCACHED_NOTSTORED\fP is a legitmate error in the case of a collision.
.SH HOME
.sp
To find out more information please check:
\fI\%http://libmemcached.org/\fP
.SH SEE ALSO
.sp
\fImemcached(1)\fP \fIlibmemached(3)\fP \fImemcached_strerror(3)\fP \fImemcached_prepend(3)\fP \fImemcached_cas(3)\fP \fImemcached_append(3)\fP
.SH AUTHOR
Brian Aker
.SH COPYRIGHT
2011-2013, Brian Aker DataDifferential, http://datadifferential.com/
.\" Generated by docutils manpage writer.
.


Changes
=head1 NAME

Memcached::libmemcached::Changes - List of significant changes

=head2 Changes in 1.001702

  Improve handling of case where last_return==MEMCACHED_ERRNO
    but lmc_state->last_errno==0, RT#41299.

=head2 Changes in 1.001701 - 2nd Sept 2013

  Sync with libmemcached 1.0.17
  Fixed links in docs, RT#66291
  Fixed whatis in pod of Memcached::libmemcached::API, RT#71172
  Fixed undef warnings related to expiration, RT#69996
  Don't let DESTDIR env var upset libmemcached build (Chip Salzenberg).

  Note that get_server_for_key() may have a memory leak.

=head2 Changes in 0.4406 4th Jan 2011

  Added get_server_for_key() thanks to Roy Hooper.

=head2 Changes in 0.4405 29th Oct 2010

  Fix crash if memcached_callback_get(MEMCACHED_CALLBACK_PREFIX_KEY) returns NULL.

=head2 Changes in 0.4404 28th Oct 2010

  Added memcached_callback_set() to enable MEMCACHED_CALLBACK_PREFIX_KEY.
  Added memcached_callback_get()
  Made test timing more robust.

=head2 Changes in 0.4403 28th Oct 2010

  Changed memcached_version() to return the lowest server version seen.
  Added new memcached_{inc,dec}rement{_with_initial,}{_by_key,} methods.
  Added new methods memcached_server_add_with_weight() and
    memcached_server_add_unix_socket_with_weight().
  Added new trace_level() method.

=head2 Changes in 0.4402 26th Oct 2010

  Changed walk_stats() API. Now uses the memcached_stat_execute() function
    and no longer passes $stats_arg as the forth argument to the callback.
  Minor updates to docs.

=head2 Changes in 0.4401 21st Oct 2010

  Sync with libmemcached 0.44
  Fixed core dump in stats function.
  Fixed memory leak in stats function. RT#41496
  Improved error reporting in trace log.

=head2 Changes in 0.4202  2nd Aug 2010

  Fix memcached_version() to properly return a list when called in list context
     this should resolve rt #59864

=head2 Changes in 0.4201 16th Jul 2010

  Sync with libmemcached 0.42
  Squashed various compile warnings
  Many updates to libmemcached API calls due to API changes
  Merged existing svn tree (which was out of sync) with 
  Daisuke's tree

=head2 Changes in 0.3101 6th Sep 2009

  Sync with libmemcached 0.31
  Now tested against memcached 1.4.1
  Now tested with binary protocol
  Work with newer libmemcached, which returned MEMCACHED_END instead of
    MEMCACHED_SUCCESS on fetch success

=head2 Changes in 0.2501 (svn r163) 17th Dec 2008

  Sync with libmemcached 0.25.

=head2 Changes in 0.2101 (svn r156) 27th May 2008

  Sync with libmemcached 0.21.
  Report memcached version when running tests.
  Skip flags test when using memcached <1.2.4.
  Test consistent hashing thanks to Daisuke Maki.
  Deprecated undocumented calling get() with an array ref as key
    it now generates a warning.
  Added way to specify multiple memcached servers for tests.
  Added -bin option to Makefile.PL to install mem* utilities
    thanks to Jos Boumans.

=head2 Changes in 0.1902 (svn r144) 17th April 2008

  Fixed leak of $memc which prevented DESTROY (and thus memcached_free)
    being called, thanks to Jim Spath and Daisuke Maki.
  Fixed tests to not use whitespace in keys, thanks to Daisuke Maki.

=head2 Changes in 0.1901 (svn r140) 10th April 2008

  Sync with libmemcached 0.19.
  Added walk_stats method for fetching stats via a callback
    thanks to Daisuke Maki.

=head2 Changes in 0.1701 (svn r129) 3rd March 2008

  Sync with libmemcached 0.17.
  Perl API for memcached_version returns the version info.
  Added auto-generated Memcached::libmemcached::constants docs.
  Improved docs.
  Assorted minor fixes.

=head2 Changes in 0.1405 (svn rXXX)

  Added official object-oriented API for perl-specific extensions.
  Allow SubclassName->new as alias for memcached_create.
  Deprecated perl-specific functions superceeded by methods:
    memcached_errstr                - now $memc->errstr
    memcached_mget_into_hashref     - now $memc->mget_into_hashref
    memcached_set_callback_coderefs - now $memc->set_callback_coderefs
  Added -g/-pg options to Makefile.PL to enable debug/profiling.
  Changed memcache object to be blessed hash instead of blessed int.

=head2 Changes in 0.1404 (svn r106) 29th Jan 2008

  Fixes for perl 5.6.
  Removed dependency on Time::HiRes for testing.

=head2 Changes in 0.1402 (svn r102) 27th Jan 2008

  Added tracing via PERL_LIBMEMCACHED_TRACE env var.
  Added perl callbacks for get and set.
  Added many more functions and improved the docs.
  Reenable perl 5.006 as we supply our own xsubpp now.

=head2 Changes in 0.1401

  Added memcached_errstr and related $memc state management.
  Changes functions to return simple boolean code.
  Added memcached_replace, memcached_cas
  More tests.

=head2 Changes in 0.1308

  Includes libmemcached-0.14
  memcached_return values are now boolean (true, defined false, or undef)
  Added memcached_errstr($memc) to access last error code and string.
  Bundle xsubpp to avoid problems interfacing with our custom ParseXS.

=head2 Changes in 0.1307

  Min perl version now 5.008 due to XS length(other_arg) syntax.

=cut



lib/Memcached/libmemcached/memcached_set.pm
package Memcached::libmemcached::memcached_set;

=head1 NAME

memcached_set, memcached_add, memcached_replace

=head1 LIBRARY

C Client Library for memcached (libmemcached, -lmemcached)

=head1 SYNOPSIS

  #include <memcached.h>

  memcached_return
    memcached_set (memcached_st *ptr,
                   const char *key, size_t key_length, 
                   const char *value, size_t value_length, 
                   time_t expiration,
                   uint32_t flags);

  memcached_return
    memcached_add (memcached_st *ptr,
                   const char *key, size_t key_length,
                   const char *value, size_t value_length, 
                   time_t expiration,
                   uint32_t flags);

  memcached_return
    memcached_replace (memcached_st *ptr,
                       const char *key, size_t key_length,
                       const char *value, size_t value_length, 
                       time_t expiration,
                       uint32_t flags);

  memcached_return 
    memcached_prepend(memcached_st *ptr, 
                      const char *key, size_t key_length,
                      const char *value, size_t value_length, 
                      time_t expiration,
                      uint32_t flags)

  memcached_return 
    memcached_append(memcached_st *ptr, 
                     const char *key, size_t key_length,
                      const char *value, size_t value_length, 
                      time_t expiration,
                      uint32_t flags)
  memcached_return 
    memcached_cas(memcached_st *ptr, 
                  const char *key, size_t key_length,
                  const char *value, size_t value_length, 
                  time_t expiration,
                  uint32_t flags,
                  uint64_t cas);

  memcached_return 
    memcached_set_by_key(memcached_st *ptr, 
                         const char *master_key, size_t master_key_length, 
                         const char *key, size_t key_length, 
                         const char *value, size_t value_length, 
                         time_t expiration,
                         uint32_t flags);

  memcached_return 
    memcached_add_by_key(memcached_st *ptr, 
                         const char *master_key, size_t master_key_length,
                         const char *key, size_t key_length,
                         const char *value, size_t value_length, 
                         time_t expiration,
                         uint32_t flags);

  memcached_return 
    memcached_replace_by_key(memcached_st *ptr, 
                             const char *master_key, size_t master_key_length,
                             const char *key, size_t key_length,
                             const char *value, size_t value_length, 
                             time_t expiration,
                             uint32_t flags);

  memcached_return 
    memcached_prepend_by_key(memcached_st *ptr, 
                             const char *master_key, size_t master_key_length,
                             const char *key, size_t key_length,
                             const char *value, size_t value_length, 
                             time_t expiration,
                             uint32_t flags);

  memcached_return 
    memcached_append_by_key(memcached_st *ptr, 
                            const char *master_key, size_t master_key_length,
                            const char *key, size_t key_length,
                            const char *value, size_t value_length, 
                            time_t expiration,
                            uint32_t flags);

  memcached_return 
    memcached_cas_by_key(memcached_st *ptr, 
                         const char *master_key, size_t master_key_length,
                         const char *key, size_t key_length,
                         const char *value, size_t value_length, 
                         time_t expiration,
                         uint32_t flags,
                         uint64_t cas);

=head1 DESCRIPTION

memcached_set(), memcached_add(), and memcached_replace() are all used to
store information on the server. All methods take a key, and its length to
store the object. Keys are currently limited to 250 characters by the
memcached(1) server. You must also supply a value and a length. Optionally you
may support an expiration time for the object and a 16 byte value (it is
meant to be used as a bitmap).

memcached_set() will write an object to the server. If an object already
exists it will overwrite what is in the server. If the object does not exist
it will be written. If you are using the non-blocking mode this function
will always return true unless a network error occurs. 

memcached_replace() replaces an object on the server. If the object is not
found on the server an error occurs.

memcached_add() adds an object to the server. If the object is found on the
server an error occurs, otherwise the value is stored.

memcached_prepend() places a segment of data before the last piece of data 
stored. Currently expiration and key are not used in the server.

memcached_append() places a segment of data at the end of the last piece of 
data stored. Currently expiration and key are not used in the server.

memcached_cas() overwrites data in the server as long as the "cas" value is 
still the same in the server. You can get the cas value of a result by 
calling memcached_result_cas() on a memcached_result_st(3) structure. At the point 
that this note was written cas is still buggy in memached. Turning on support
for it in libmemcached(3) is optional. Please see memcached_set() for 
information on how to do this.

memcached_set_by_key(), memcached_add_by_key(), memcached_replace_by_key(), 
memcached_prepend_by_key(), memcached_append_by_key_by_key(), 
memcached_cas_by_key() methods all behave in a similar method as the non key 
methods. The difference is that they use their master_key parameter to map
objects to particular servers.

If you are looking for performance, memcached_set() with non-blocking IO is 
the fastest way to store data on the server.

=head1 RETURN

All methods return a value of type C<memcached_return>.
On success the value will be C<MEMCACHED_SUCCESS>.
Use memcached_strerror() to translate this value to a printable string.

For memcached_replace() and memcached_add(), C<MEMCACHED_NOTSTORED> is a
legitmate error in the case of a collision.

=head1 HOME

To find out more information please check:
L<http://tangent.org/552/libmemcached.html>

=head1 AUTHOR

Brian Aker, E<lt>brian@tangent.orgE<gt>

=head1 SEE ALSO

memcached(1) libmemached(3) memcached_strerror(3)

=cut

1;
